src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,367 @@
+@file:Suppress("MemberVisibilityCanBePrivate")
+
+package app.revanced.patcher.apk
+
+import app.revanced.patcher.DomFileEditor
+import app.revanced.patcher.Patcher
+import app.revanced.patcher.PatcherOptions
+import app.revanced.patcher.resource.*
+import app.revanced.patcher.util.ProxyBackedClassList
+import app.revanced.patcher.util.xml.LazyXMLInputSource
+import com.reandroid.apk.ApkModule
+import com.reandroid.apk.xmlencoder.EncodeException
+import com.reandroid.apk.xmlencoder.EncodeMaterials
+import com.reandroid.apk.xmlencoder.EncodeUtil
+import com.reandroid.archive.InputSource
+import com.reandroid.arsc.chunk.PackageBlock
+import com.reandroid.arsc.chunk.TableBlock
+import com.reandroid.arsc.chunk.xml.AndroidManifestBlock
+import com.reandroid.arsc.decoder.ValueDecoder
+import com.reandroid.arsc.value.Entry
+import com.reandroid.arsc.value.ResConfig
+import com.reandroid.xml.XMLDocument
+import com.reandroid.xml.XMLSpannable
+import lanchon.multidexlib2.BasicDexEntry
+import lanchon.multidexlib2.DexIO
+import lanchon.multidexlib2.MultiDexContainerBackedDexFile
+import lanchon.multidexlib2.MultiDexIO
+import lanchon.multidexlib2.RawDexIO
+import org.jf.dexlib2.dexbacked.DexBackedDexFile
+import org.jf.dexlib2.iface.DexFile
+import org.jf.dexlib2.iface.MultiDexContainer
+import org.jf.dexlib2.writer.io.MemoryDataStore
+import java.io.File
+import java.io.InputStream
+import java.util.zip.ZipEntry
+
+
+/**
+ * An [Apk] file.
+ */
+sealed class Apk private constructor(internal val module: ApkModule) {
+    companion object {
+        const val manifest = "AndroidManifest.xml"
+    }
+
+    /**
+     * The metadata of the [Apk].
+     */
+    val packageMetadata = PackageMetadata(module.androidManifestBlock)
+
+    internal val archive = Archive(module)
+
+    /**
+     * Update the [PackageBlock] name to match the manifest.
+     */
+    protected open fun updatePackageBlock() {
+        resources.packageBlock!!.name = module.apkArchive.getInputSource(manifest).openStream()
+            .use { stream -> AndroidManifestBlock.load(stream) }.packageName
+    }
```

Comment:
I feel like this doesn't belong here. Not every `Apk`has a package block, such as a library apk.

Reply:
Hmm... library apks are an exception to the rule , which means inheritance kinda falls apart.

I'm not sure what is the best way to deal with this.